core/src/main/java/org/infinispan/container/impl/EntryFactoryImpl.java

```diff
@@ -62,34 +66,108 @@ public void init() {
    }
 
    @Override
-   public final void wrapEntryForReading(InvocationContext ctx, Object key, int segment, boolean isOwner) {
+   public final CompletionStage<Void> wrapEntryForReading(InvocationContext ctx, Object key, int segment, boolean isOwner) {
       if (!isOwner && !isL1Enabled) {
-         return;
+         return CompletableFutures.completedNull();
       }
+      CompletionStage<Void> returnedStage = CompletableFutures.completedNull();
       CacheEntry cacheEntry = getFromContext(ctx, key);
       if (cacheEntry == null) {
-         cacheEntry = getFromContainerForRead(key, segment, isOwner);
+         InternalCacheEntry readEntry = getFromContainer(key, segment);
+         CompletionStage<Boolean> expiredStage = null;
+         if (readEntry == null) {
+            if (isOwner) {
+               addReadEntryToContext(ctx, NullCacheEntry.getInstance(), key);
+            }
+         } else if (isOwner || readEntry.isL1Entry()) {
+            cacheEntry = readEntry;
+            expiredStage = handlePossibleExpiredEntry(readEntry, false);
 
-         if (cacheEntry != null) {
-            // With repeatable read, we need to create a RepeatableReadEntry as internal cache entries are mutable
-            // Otherwise we can store the InternalCacheEntry directly in the context
-            if (useRepeatableRead) {
-               MVCCEntry mvccEntry = createWrappedEntry(key, cacheEntry);
-               mvccEntry.setRead();
-               cacheEntry = mvccEntry;
+            if (expiredStage != null) {
+               if (CompletionStages.isCompletedSuccessfully(expiredStage)) {
+                  Boolean expired = CompletionStages.join(expiredStage);
+                  if (expired == Boolean.FALSE) {
```

Comment:
nitpick: move this code block to a private method? 
avoids duplication in the lambda below.

Reply:
Refactored